### ISSUE TITLE
[IX] Support manual selection of training sample

### ISSUE THREAD

**Issue by aphilop** [huridocs/uwazidocs#8166]
### Problem statement

With the currently implemented flow, a user can extract first paragraphs through PX and they get automatically as title <document_title.paragraph_number>. Then, if the user wants to change the paragraph title through metadata extraction, a relevant extractor does not work in practice because all entities have already a value (the title <document_title.paragraph_number>) and even after some training iterations it can not provide the expected results. This limitation can significantly impact certain partners that would want the extracted paragraphs to have different titles.

### Proposed solution

- Create a manual training sample curation system that allows users to select which entries should be used for training metadata extractors
- Build mechanisms for bulk selection, filtering, and management of entities for the training sample
- Ensure training samples persist across multiple training sessions
- Provide visibility into which entities are being used for training

### Acceptance criteria

- [ ] Users can individually mark/unmark entities as "use for training"
- [ ] Users can select multiple entities at once using a bulk action for setting the training sample
- [ ] ~When an entity is manually marked for training, empty values are accepted as valid training data~ -> not to be considered for the moment
- [ ] Training samples (per extractor) should persist across training iterations
- [ ] Users must be able to switch between manual training (selected entities only) and automatic training modes (default behavior)
- [ ] Manual training mode only uses explicitly selected entities for training, ignoring all others
- [ ] Each entity clearly shows whether it's included in the training sample (badge, icon, or visual indicator)
- [ ] Provide a filter view to show only entities marked for training
- [ ] System must handle large selections (thousands of selected entities for training) without breaking

**Comment by aphilop** [comment]
Still open to define:
- how to ensure enough data are sent for training and communicate that to the user?
- If empty values are accepted as valid training data, how does this affect what we display in the filters (labeled/non-labeled)?

**Comment by txau** [comment]
Regarding the problem statement:

There are several scenarios in which being able to cherry-pick the train set is needed, or at least convenient, the title one you described being one of them, but more generally speaking having mixed training data with good samples and bad samples. Examples: 
- the title is automatically set and the format is not desired
- a change in the labeling criteria 
- training is not converging properly
- ensure edge cases are added to the training set

> When an entity is manually marked for training, empty values are accepted as valid training data

I would put this criteria on hold for now. I'll provide rationale on the relevant issue.

IMO this is good enough for a first iteration of the feature.

**Comment by txau** [comment]
I think we should take the time to carefully review this proposal, consider what is explained in this issue: https://github.com/huridocs/ml-backlog/issues/134

**Comment by txau** [comment]
I have a suggestion on how to solve this feature that may help us resolve these issues:
- https://github.com/huridocs/ml-backlog/issues/134
- https://github.com/huridocs/uwazi/issues/8281
- https://github.com/huridocs/uwazi/issues/8166

We can add a new property to suggestions, or rather, to the values used to generate the suggestions, in order to classify  them in:
- Gold standard: Data that has been labeled with the highest reliability, often by domain experts or with rigorous consensus.
- Silver standard: Labels derived from heuristic rules, weak supervision, or crowd workers without full expert oversight.
- Unlearnable:  Samples that the model cannot learn from because the features don’t provide enough signal, or because the task is ill-defined for that sample.

Gold standard is manually labeled as such.
Silver is the default state of any non-empty value.
Unlearnable is also manually labeled.

How does it work

Use case: some data has value (silver standard) but it is not the desired value. E.g. all entities already have a title but it is not following the desired title pattern. In this case, the user can manually create a gold standard set, train with that set alone and get the suggestions based on that training set.

Use case: a user finds a case that is mainly noise and does not offer features that allow inference. Manually adding "unlearnable" will exclude this sample from the training set reducing the amount of noise during training.

Use case: an empty string is the desired value. By labeling it as gold standard, the user is signaling the desire to train the model to return an empty string as the correct value.

Upon training, the user is present with a dialog to train with only the gold standard set, or also include the silver. Unlearnable and empty silver is always excluded.

This approach needs further definition, but I wanted to present the idea.

The naming comes from standard ML literature, but we can turn it into a more use friendly terminology if we believe it provides a better UX:
- Human Reviewed (gold standard)
- Machine generated (silver)
- Invalid sample (unlearnable) 

cc @RafaPolit @gabriel-piles let me know your thoughts

edit:

Another desirable side effect of this approach is being able to control the size of the training set. E.g. image we have a collection with hundreds of thousands of suggestions, after training for a while and accepting all the suggestions, with our current approach, all accepted suggestions become training data. Meaning, next time we want to train, it will use all the 100.000+ samples. With a gold-standard, we can tweak and keep the training set smaller.

Yet another desirable outcome is balancing out the train set. E.g. for a multi-class classifier, making sure that I have a good representation of less common and under-represented cases.

On a final note: this feature can be greatly simplified by just adding a property "Use for training" to suggestions/values, which covers all the cases: If it is labeled as use for training, it means it is gold standard, therefor it is not low quality empty silver or unlearnable.

I would definitely separated the train and find suggestions, and replace them with these buttons:
- Train -> will only use the "Use for training" samples
- Get suggestions -> if some are selected, get those, if none pop up a dialog with a selector for how many (default 1000, option for all). A third option for getting the suggestions only for un-processed (new entities). Maybe an option to get suggestions for mismatches (e.g. a new model has been trained and I want to see if previous failing suggestions now are coming back correctly).

**Comment by gabriel-piles** [comment]
@txau I like this idea; it solves several issues at once.

**Comment by txau** [comment]
Closing in favor of: https://github.com/huridocs/uwazi/issues/8408
